src/search.rs

```diff
@@ -94,6 +94,118 @@ pub enum Selectors<T> {
 
 type AttributeToCrop<'a> = (&'a str, Option<usize>);
 
+/// The query filter. This enum uses the `One` variant to represent single strings and the `Many`
+/// variant to represent an array of filters. Acceptable filter forms are described in the
+/// [dedicated guide](https://docs.meilisearch.com/reference/features/filtering.html).
+///
+/// # Examples
+///
+/// - A single filter which must hold true
+/// ```
+/// # use std::borrow::Cow;
+/// # use meilisearch_sdk::search::Filter;
+/// Filter::One("kind = \"text\"".into()); // "kind = \"text\""
+/// ```
+///
+/// - Multiple filters were all must hold true
+/// ```
+/// # use std::borrow::Cow;
+/// # use meilisearch_sdk::search::Filter;
+/// let timestamp = 1639200000;
+/// Filter::Many(vec![
+///     Filter::One("kind = \"text\"".into()),
+///     Filter::One(format!("createdAt > {}", timestamp).into())
+/// ]); // [ "kind = \"text\"", "createdAt > 1639200000" ]
+/// ```
+///
+/// - Multiple filters where any of the inner filter strings must hold true and all of the outer
+///   filters must hold true
+/// ```
+/// # use std::borrow::Cow;
+/// # use meilisearch_sdk::search::Filter;
+/// # let timestamp = 1639200000;
+/// Filter::Many(vec![
+///     Filter::Many(vec![
+///         Filter::One("kind = \"text\"".into()),
+///         Filter::One("kind = \"title\"".into())
+///     ]),
+///     Filter::Many(vec![
+///         Filter::One(format!("createdAt > {}", timestamp).into())
+///     ])
+/// ]); // [ [ "kind = \"text\"", "kind = \"title\"" ], [ "createdAt > 1639200000" ] ]
+/// ```
+///
+/// - For convenience, a few `From` conversions are also provided:
+/// ```
+/// # use std::borrow::Cow;
+/// # use meilisearch_sdk::search::Filter;
+/// # let timestamp = 1639200000;
+/// // From &str:
+/// assert_eq!(
+///     Filter::from("kind = \"text\""),
+///     Filter::One(Cow::Borrowed("kind = \"text\""))
+/// );
+/// // From String
+/// assert_eq!(
+///     Filter::from(format!("createdAt > {}", timestamp)),
+///     Filter::One(Cow::Owned("createdAt > 1639200000".to_string()))
+/// );
+/// // From Vec<F> where F can be converted to a filter
+/// assert_eq!(
+///     Filter::from(vec![
+///         vec![ Cow::from("kind = \"text\""), Cow::from("kind = \"title\"") ],
+///         vec![ Cow::from(format!("createdAt > {}", timestamp)) ]
+///     ]),
+///     Filter::Many(vec![
+///         Filter::Many(vec![
+///             Filter::One(Cow::Borrowed("kind = \"text\"")),
+///             Filter::One(Cow::Borrowed("kind = \"title\""))
+///         ]),
+///         Filter::Many(vec![
+///             Filter::One(Cow::Owned("createdAt > 1639200000".to_string()))
+///         ])
+///     ])
+/// );
+/// ```
+#[derive(Debug, Serialize, Clone, PartialEq)]
+#[serde(untagged)]
+pub enum Filter<'a> {
+    /// A single filter string
+    One(Cow<'a, str>),
+    /// Multiple filters represented by an array
+    Many(Vec<Filter<'a>>),
+}
```

Comment:
I am not sure about this type, I mean, it is small and pretty **but** it is not the real type that MeiliSearch accepts. MeiliSearch only accepts direct strings (i.e. `"count = 0"`), arrays of strings (i.e. `["count = 0", "value < 3"]`) or, arrays or arrays of strings mixed (i.e. `["value > 32", ["count = 0", "price > 50"]]`).

I am not sure about whether we should fully and correctly type what MeiliSearch accepts but it has an advantage: you are sure about the syntax you use and you cannot receive an error about the fact that you sent an array of array of array of 4 depths (i.e. `[[[["count = 0", "price > 50"]]]]`). The current type will let you express this kind of deep recursive type. Fixing that would probably involve depending on [the `either` crate](https://docs.rs/either) and will complexity the usage of the library a little more.

In another hand, this Rust SDK doesn't check the string syntax at all and lets the user write everything he wants in the strings, letting MeiliSearch answer with an error if the syntax is invalid (i.e. "hello - &what"). So I would say that we can keep this new implementation you've just done like that and let MeiliSearch validate the filters format and syntax like before.

Reply:
Yeah, I was thinking about that but exactly because there's no checking of the string format here, I don't think it's worth checking the Filter format either - Especially if the user wants to implement custom filters, the current type would help a lot more.